src/Symfony/Component/HttpClient/CachingHttpClient.php

```diff
@@ -11,68 +11,366 @@
+            unset($defaultOptions['allow_revalidate']);
+            unset($defaultOptions['stale_while_revalidate']);
+            unset($defaultOptions['stale_if_error']);
+            unset($defaultOptions['trace_level']);
+            unset($defaultOptions['trace_header']);
+        } else {
+            $this->cache = $cache;
+        }
 
         if ($defaultOptions) {
             [, $this->defaultOptions] = self::prepareRequest(null, null, $defaultOptions, $this->defaultOptions);
         }
     }
 
     public function request(string $method, string $url, array $options = []): ResponseInterface
+    {
+        if ($this->cache instanceof HttpCache) {
+            return $this->legacyRequest($method, $url, $options);
+        }
+
+        [$fullUrl, $options] = self::prepareRequest($method, $url, $options, $this->defaultOptions);
+
+        $fullUrl = implode('', $fullUrl);
+        $fullUrlTag = hash('xxh3', $fullUrl);
+
+        if (\in_array($method, self::UNSAFE_METHODS, true)) {
+            $this->cache->invalidateTags([$fullUrlTag]);
+        }
+
+        if ($options['body'] !== '' || isset($options['normalized_headers']['range']) || !\in_array($method, self::CACHEABLE_METHODS, true)) {
+            return new AsyncResponse($this->client, $method, $url, $options);
+        }
+
+        $requestHash = $this->getRequestHash($method, $fullUrl);
+        $varyKey = "vary_{$requestHash}";
+        $varyItem = $this->cache->getItem($varyKey);
+        $varyFields = $varyItem->isHit() ? $varyItem->get() : [];
+
+        $metadataKey = $this->getMetadataKey($requestHash, $options['normalized_headers'], $varyFields);
+        $metadataItem = $this->cache->getItem($metadataKey);
+        $cachedData = $metadataItem->isHit() ? $metadataItem->get() : null;
+
+        $freshness = null;
+        if (\is_array($cachedData)) {
+            $freshness = $this->evaluateCacheFreshness($cachedData);
+
+            if (Freshness::Fresh === $freshness) {
+                return $this->createResponseFromCache($metadataKey, $cachedData, $method, $url, $options, $fullUrlTag);
+            }
+
+            if (isset($cachedData['headers']['etag'])) {
+                $options['headers']['If-None-Match'] = implode(', ', $cachedData['headers']['etag']);
+            }
+
+            if (isset($cachedData['headers']['last-modified'][0])) {
+                $options['headers']['If-Modified-Since'] = $cachedData['headers']['last-modified'][0];
+            }
+        }
+
+        $chunkIndex = -1;
+        // consistent expiration time for all items
+        $expiresAt = null === $this->maxTtl
+            ? null
+            : \DateTimeImmutable::createFromFormat('U', time() + $this->maxTtl);
+
+        return new AsyncResponse(
+            $this->client,
+            $method,
+            $url,
+            $options,
+            function (ChunkInterface $chunk, AsyncContext $context) use (
+                &$chunkIndex,
```

Comment:
I think some vars used by reference can be turned into static variables inside the closure. That might be a bit cleaner to me.